backoff.c

```diff
@@ -0,0 +1,87 @@
+#include "common.h"
+
+#include <ext/standard/php_rand.h>
+#include <ext/standard/php_mt_rand.h>
+
+#include "backoff.h"
+
+zend_long random_range(zend_long min, zend_long max) {
+    if (max <= min) {
+        max = min + 1;
+    }
+
+  return php_mt_rand_range(min, max);
+}
+
+zend_long default_backoff(struct Backoff *self, int retry_index) {
+    zend_long backoff = retry_index ? self->base : random_range(0, self->base);
+    return MIN(self->cap, backoff);
+}
+
+zend_long constant_backoff(struct Backoff *self, int retry_index) {
+    zend_long backoff = self->base;
+    return MIN(self->cap, backoff);
```

Comment:
Perhaps this can be simplified a bit

```suggestion
    return MIN(self->base, self->cap);
```

Reply:
I wrote it that way to have the exact same return in all implementations, to make it clear. But I'm open to changing it.